codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/codegen/ValidationGenerator.java

```diff
@@ -109,91 +172,179 @@ private void renderValidatorForEachShape(
-    }
-    if (currentShape.hasTrait(DafnyUtf8BytesTrait.class)) {
-      addUTFCheck(currentShape, dataSource, pointableString);
-    }
+    validationCode.append(
+      addRangeCheck(memberShape, dataSource, pointableString)
+    );
+    validationCode.append(
+      addLengthCheck(memberShape, dataSource, pointableString)
+    );
+    validationCode.append(addRequiredCheck(symbol, memberShape, dataSource));
+    validationCode.append(
+      addUTFCheck(memberShape, dataSource, pointableString)
+    );
     // Broke list and map into two different if else because for _, item := range %s looked good for list
     // And for key, value := range %s looked good for map
     if (currentShape.isListShape()) {
-      writer.write(
-        """
-        for _, %s := range %s {
-            // To avoid declared and not used error for shapes which does not need validation check
-            _ = item
-        """.formatted(LIST_ITEM, dataSource)
-      );
-      renderValidatorHelper(currentShape, false, LIST_ITEM);
-      writer.write(
-        """
+      MemberShape itemMember = ((ListShape) currentShape).getAllMembers().values().iterator().next();
+      StringBuilder itemValidation = new StringBuilder();
+      renderValidatorForEachShape(model.expectShape(itemMember.getTarget()), itemMember, false, LIST_ITEM, itemValidation);
+      // If the validation function is not created and the list shape does have some constraints
+      if (!validationFuncMap.containsKey(memberShape) && itemValidation.length() != 0) {
+        final String funcName = funcNameGenerator(memberShape, "validate");
+        final String funcInput = dataSource.startsWith("input") ? "" : dataSource;
+        if (!funcInput.equals("")) {
+          String inputType = GoCodegenUtils.getType(symbolProvider.toSymbol(currentShape), currentShape);
+          // remove the package name because this code is generated inside smithyTypesNamespace itself
+          inputType =
+            inputType.replace(
+              SmithyNameResolver.smithyTypesNamespace(currentShape).concat("."),
+              ""
+            );
+          validationFuncInputTypeMap.put(memberShape, inputType);
+          dataSource = "Value";
         }
-        """
-      );
+        String funcCall = "input.".concat(funcName).concat("(%s)".formatted(funcInput));
+        validationCode.append(
+          CHECK_AND_RETURN_ERROR.formatted(
+            funcCall, funcCall
+          )
+        );
+        validationFuncMap.put(memberShape, null);
+        StringBuilder listValidation = new StringBuilder();
+        listValidation.append(
+          """
+          for _, %s := range %s {
+          """.formatted(LIST_ITEM, dataSource)
+        );
+        listValidation.append(itemValidation);
+        listValidation.append(
+          """
+          }
+          """
+        );
+        validationFuncMap.put(memberShape, listValidation.toString());
+      }
     } else if (currentShape.isMapShape()) {
```

Comment:
Use switch case with delegation to methods for each shape.

Reply:
Done